fsi/mapping.go

```diff
@@ -1,436 +1,122 @@
 package fsi
 
 import (
-	"encoding/json"
 	"fmt"
-	"io"
-	"io/ioutil"
 	"os"
-	"path/filepath"
 	"strings"
-	"time"
 
-	"github.com/ghodss/yaml"
 	"github.com/qri-io/dataset"
-	"github.com/qri-io/dataset/detect"
-	"github.com/qri-io/dataset/dsio"
+	"github.com/qri-io/qri/fsi/component"
 )
 
-const (
-	componentNameCommit    = "commit"
-	componentNameDataset   = "dataset"
-	componentNameMeta      = "meta"
-	componentNameBody      = "body"
-	componentNameStructure = "structure"
-	componentNameTransform = "transform"
-	componentNameViz       = "viz"
-)
-
-var (
-	// ErrNoDatasetFiles indicates no data
-	ErrNoDatasetFiles = fmt.Errorf("no dataset files provided")
-)
-
-// FileStat holds information about a file: its path and mod time
-type FileStat struct {
-	Path  string
-	Mtime time.Time
-}
-
-// ReadDir parses a directory into a dataset, returning both the dataset and
-// a map of component names to the files they came from. Files can be specified
-// in either JSON or YAML format. It is an error to specify any component more
-// than once
-func ReadDir(dir string) (ds *dataset.Dataset, fileMap, problems map[string]FileStat, err error) {
-	fileMap = map[string]FileStat{}
-	ds = &dataset.Dataset{}
-	problems = nil
-
-	components := map[string]interface{}{
-		componentNameDataset: ds,
-
-		componentNameCommit:    &dataset.Commit{},
-		componentNameMeta:      &dataset.Meta{},
-		componentNameStructure: &dataset.Structure{},
-		componentNameTransform: &dataset.Transform{},
-		componentNameViz:       &dataset.Viz{},
-
-		componentNameBody: nil,
-	}
-
-	extensions := map[string]decoderFactory{
-		".json": newJSONDecoder,
-		".yaml": newYAMLDecoder,
-		".yml":  newYAMLDecoder,
+// ReadDir reads the component files in the directory, and returns a dataset
+func ReadDir(dir string) (*dataset.Dataset, error) {
```

Comment:
lit. I'd very much like to see this accept a context. We've encountered some situations where reading from FSI slowness has choked holding an os-level file descriptor. Accepting a context here sets the stage for adding a deadline later